presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentProvider.java

```diff
@@ -45,14 +46,19 @@ protected EnvironmentProvider(List<EnvironmentExtender> bases)
                 .add(environmentConfig)
                 .build();
 
-        String extendersNames = extenders.stream()
-                .map(Object::getClass)
-                .map(Class::getSimpleName)
-                .collect(joining(", "));
-
-        log.info("Building environment %s with extenders: %s", builder.getEnvironmentName(), extendersNames);
-
-        extenders.forEach(extender -> extender.extendEnvironment(builder));
+        Set<Class<? extends EnvironmentExtender>> seen = new HashSet<>();
+        extenders.forEach(extender -> extend(extender, builder, seen));
         return builder;
     }
+
+    private void extend(EnvironmentExtender extender, Environment.Builder builder, Set<Class<? extends EnvironmentExtender>> seen)
+    {
+        extender.getDependencies()
+                .forEach(dependencyExtender -> extend(dependencyExtender, builder, seen));
+        Class<? extends EnvironmentExtender> extenderClass = extender.getClass();
+        if (seen.add(extenderClass)) {
```

Comment:
I don't like that. Nothing prevents the extenders to be parametrized. We pass instances everywhere. And here we are filtering based on class.

Reply:
👍 

I will use identity set instead and use regular objects.